### PR TITLE
Update fill fe values argument list

### LINKED ISSUE #1277: FEValues cleanup: resort arguments of fill_fe_values and document them
For #1198, addresses the last still open point of #1238.

### PR DESCRIPTION
**Pull request by bangerth**
As discussed in #1277, the list of arguments of `FiniteElement::fill_fe_*_values` is a bit "Kraut und Rueben". It needed to be sorted in a somewhat meaningful way, and the arguments needed to be documented appropriately.

This patch does this. The first of the two commits just reorders arguments (and standardizes on a common naming scheme for the arguments in all implementations). It is very boring. The second commit provides documentation of these functions and their arguments.

This fixes #1277.

### CODE REVIEW

**Comment by bangerth** [comment]
In reference to #1198.

File: include/deal.II/fe/fe.h [hunk 1]
```diff
@@ -2248,44 +2248,219 @@ class FiniteElement : public Subscriptor,
...
+   *   a reference to this object to certain functions of the
+   *   Mapping class (e.g., Mapping::transform()) if you need to
+   *   call them from the current function.
+   * @param[in] mapping_data The output object into which the
+   *   Mapping::fill_fe_values() function wrote the mapping information
+   *   corresponding to the current cell. This includes, for example,
+   *   Jacobians of the mapping that may be of relevance to the
+   *   current function, as well as other information that FEValues::reinit()
+   *   requested from the mapping.
+   * @param[in] fe_internal_ A reference to an object previously
```
Comment (kronbichler) [inline include/deal.II/fe/fe.h]: fe_internal_ -> fe_internal

File: include/deal.II/fe/fe.h [hunk 2]
```diff
@@ -2248,44 +2248,219 @@ class FiniteElement : public Subscriptor,
...
+   *   is to compute a mapping from the reference cell to.
+   * @param[in] cell_similarity Whether or not the cell given as first
+   *   argument is simply a translation, rotation, etc of the cell for
+   *   which this function was called the most recent time. This
+   *   information is computed simply by matching the vertices (as stored
+   *   by the Triangulation) between the previous and the current cell.
+   *   The value passed here may be modified by implementations of
+   *   this function and should then be returned (see the discussion of the
+   *   return value of this function).
+   * @param[in] quadrature A reference to the quadrature formula in use
```
Comment (kronbichler) [inline include/deal.II/fe/fe.h]: I was surprised that we pass a quadrature object to fill_fe_values. I just had a short look at the respective methods fe_poly.templates.h, fe_poly_tensor.cc, fe_face.cc and I couldn't find a proper use of 'quadrature' - it seems that all we use it for is to find out the loop lengths in various places. But that we should be able to deduce from the mapping related data (if necessary, we can also include a new variable). Or do I miss a place where we actually need the quadrature object (I didn't look too carefully)?

File: include/deal.II/fe/fe.h [hunk 3]
```diff
@@ -2248,44 +2248,219 @@ class FiniteElement : public Subscriptor,
...
+   *   a reference to this object to certain functions of the
+   *   Mapping class (e.g., Mapping::transform()) if you need to
+   *   call them from the current function.
+   * @param[in] mapping_data The output object into which the
+   *   Mapping::fill_fe_values() function wrote the mapping information
+   *   corresponding to the current cell. This includes, for example,
+   *   Jacobians of the mapping that may be of relevance to the
+   *   current function, as well as other information that FEValues::reinit()
+   *   requested from the mapping.
+   * @param[in] fe_internal_ A reference to an object previously
```
Comment (kronbichler) [inline include/deal.II/fe/fe.h]: same here

File: include/deal.II/fe/fe.h [hunk 4]
```diff
@@ -2248,44 +2248,219 @@ class FiniteElement : public Subscriptor,
...
```
Comment (kronbichler) [inline include/deal.II/fe/fe.h]: here too

**Comment by kronbichler** [comment]
Apart from the comment regarding the 'quadrature' argument and the typos, this looks good.

**Comment by bangerth** [comment]
It's a good question about the quadrature. `Mapping::fill_fe_values` also gets it, and I think there it makes sense (because the mappings have to compute the JxW values) so I suppose when we wrote this, the argument was just copied.

I don't feel particularly attached to it. If an element really needs it, it could as well copy it into its `InternalData` class. On the other hand, saying `quadrature.size()` is more expressive than getting the number of points from one of the other arrays that just happen to have that size. The cost of passing a reference to such an expensive function is also rather minimal.

I've updated the typos (thanks!). Let me know what you'd like me to do about the quadrature.

**Comment by kronbichler** [comment]
I do not feel overly strong about changing it. I appreciate concise interfaces and getting rid of one argument seems good. On the other hand, the arguments are a mess anyway so I don't think it's worth to check more closely whether it really is that simple. So I'll merge now.

**Comment by bangerth** [comment]
Yes, these functions have a lot of arguments. On the other hand, they're not used in user-visible interfaces, so I'm not overly concerned.

**Pull request merged**